flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/codegen/HashCodeGeneratorTest.scala

```diff
@@ -41,16 +45,164 @@ class HashCodeGeneratorTest {
 
     val hashFunc2 = HashCodeGenerator
       .generateRowHash(
-        new CodeGeneratorContext(new Configuration, Thread.currentThread().getContextClassLoader),
+        new CodeGeneratorContext(new Configuration, classLoader),
         RowType.of(new IntType(), new BigIntType(), new VarBinaryType(VarBinaryType.MAX_LENGTH)),
         "name",
         Array(1, 2, 0)
       )
       .newInstance(classLoader)
 
     val row = GenericRowData.of(ji(5), jl(8), Array[Byte](1, 5, 6))
-    Assert.assertEquals(637, hashFunc1.hashCode(row))
-    Assert.assertEquals(136516167, hashFunc2.hashCode(row))
+    assertEquals(637, hashFunc1.hashCode(row))
+    assertEquals(136516167, hashFunc2.hashCode(row))
+
+    // test row with nested array and map type
+    val hashFunc3 = HashCodeGenerator
+      .generateRowHash(
+        new CodeGeneratorContext(new Configuration, classLoader),
+        RowType.of(
+          new IntType(),
+          new ArrayType(new IntType()),
+          new MultisetType(new IntType()),
+          new MapType(new IntType(), new VarCharType())),
+        "name",
+        Array(1, 2, 0, 3)
+      )
+      .newInstance(classLoader)
+
+    val row3 = GenericRowData.of(
+      ji(5),
+      new GenericArrayData(Array(1, 5, 7)),
+      new GenericMapData(Map(1 -> null, 5 -> null, 10 -> null)),
+      new GenericMapData(
+        Map(1 -> StringData.fromString("ron"), 5 -> StringData.fromString("danny"), 10 -> null))
+    )
+    assertEquals(1065781729, hashFunc3.hashCode(row3))
+
+    // test hash code for ArrayData
+    CommonTestUtils.assertThrows(
+      "RowData hash function doesn't support to generate hash code for ArrayData.",
+      classOf[RuntimeException],
+      () => hashFunc3.hashCode(new GenericArrayData(Array(1)))
+    )
+
+    // test hash code for MapData
+    CommonTestUtils.assertThrows(
+      "RowData hash function doesn't support to generate hash code for MapData.",
+      classOf[RuntimeException],
+      () => hashFunc3.hashCode(new GenericMapData(null)))
+  }
+
+  @Test
+  def testArrayHash(): Unit = {
+    // test primitive type
+    val hashFunc1 = HashCodeGenerator
+      .generateArrayHash(
+        new CodeGeneratorContext(new Configuration(), classLoader),
+        new IntType(),
+        "name")
+      .newInstance(classLoader)
+
+    val array1 = new GenericArrayData(Array(1, 5, 7))
+    assertEquals(1123, hashFunc1.hashCode(array1))
```

Comment:
Asserting the hash int is hard to tell right or wrong. Would be better to add assertions, e.g. 

```
val array1 = new GenericArrayData(Array(1, 5, 7))
    val array2 = new GenericArrayData(Array(1, 5, 7))
    val array3 = new GenericArrayData(Array[AnyRef](1, null, 5, null, 7))
    assertEquals(hashFunc1.hashCode(array1), hashFunc1.hashCode(array2))
    assertNotEquals(hashFunc1.hashCode(array1), hashFunc1.hashCode(array3))
```